pkg/virt-controller/watch/workload-updater/workload-updater.go

```diff
@@ -0,0 +1,554 @@
+	kubeVirtInformer cache.SharedIndexInformer,
+	recorder record.EventRecorder,
+	clientset kubecli.KubevirtClient,
+	clusterConfig *virtconfig.ClusterConfig,
+) *WorkloadUpdateController {
+
+	c := &WorkloadUpdateController{
+		queue:                 workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+		vmiInformer:           vmiInformer,
+		migrationInformer:     migrationInformer,
+		kubeVirtInformer:      kubeVirtInformer,
+		recorder:              recorder,
+		clientset:             clientset,
+		statusUpdater:         status.NewKubeVirtStatusUpdater(clientset),
+		migrationExpectations: controller.NewUIDTrackingControllerExpectations(controller.NewControllerExpectations()),
+		clusterConfig:         clusterConfig,
+
+		throttleIntervalSeconds: defaultThrottleIntervalSeconds,
+
+		reconcileThrottleMap: make(map[string]time.Time),
+	}
+
+	c.kubeVirtInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.addKubeVirt,
+		DeleteFunc: c.deleteKubeVirt,
+		UpdateFunc: c.updateKubeVirt,
+	})
+
+	c.migrationInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.addMigration,
+		DeleteFunc: c.deleteMigration,
+		UpdateFunc: c.updateMigration,
+	})
+
+	return c
+}
+
+func (c *WorkloadUpdateController) getKubeVirtKey() (string, error) {
+	kvs := c.kubeVirtInformer.GetStore().List()
+	if len(kvs) > 1 {
+		log.Log.Errorf("More than one KubeVirt custom resource detected: %v", len(kvs))
+		return "", fmt.Errorf("more than one KubeVirt custom resource detected: %v", len(kvs))
+	}
+
+	if len(kvs) == 1 {
+		kv := kvs[0].(*virtv1.KubeVirt)
+		return controller.KeyFunc(kv)
+	}
+	return "", nil
+}
+
+func (c *WorkloadUpdateController) addMigration(obj interface{}) {
+	migration, ok := obj.(*virtv1.VirtualMachineInstanceMigration)
+	if !ok {
+		return
+	}
+
+	key, err := c.getKubeVirtKey()
+	if key == "" || err != nil {
+		return
+	}
+
+	if migration.Annotations != nil {
+		// only observe the migration expectation if our controller created it
+		_, ok = migration.Annotations[virtv1.WorkloadUpdateMigrationAnnotation]
+		if ok {
+			c.migrationExpectations.CreationObserved(key)
+		}
+	}
+
+	c.queue.Add(key)
+}
+
+func (c *WorkloadUpdateController) deleteMigration(obj interface{}) {
+	key, err := c.getKubeVirtKey()
+	if key == "" || err != nil {
+		return
+	}
+
+	c.queue.Add(key)
+}
+
+func (c *WorkloadUpdateController) updateMigration(old, curr interface{}) {
+	key, err := c.getKubeVirtKey()
+	if key == "" || err != nil {
+		return
+	}
+
+	c.queue.Add(key)
+}
+
+func (c *WorkloadUpdateController) addKubeVirt(obj interface{}) {
+	c.enqueueKubeVirt(obj)
+}
+
+func (c *WorkloadUpdateController) deleteKubeVirt(obj interface{}) {
+	c.enqueueKubeVirt(obj)
+}
+
+func (c *WorkloadUpdateController) updateKubeVirt(old, curr interface{}) {
+	c.enqueueKubeVirt(curr)
+}
+
+func (c *WorkloadUpdateController) enqueueKubeVirt(obj interface{}) {
+	logger := log.Log
+	kv := obj.(*virtv1.KubeVirt)
```

Comment:
This cast can fail under certain scenarios. In theory you  can get a tombstone object and you have to do something like this (copied from another watcher):

```
	domain, ok := obj.(*api.Domain)
	if !ok {
		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
		if !ok {
			log.Log.Reason(fmt.Errorf("couldn't get object from tombstone %+v", obj)).Error("Failed to process delete notification")
			return
		}
		domain, ok = tombstone.Obj.(*api.Domain)
		if !ok {
			log.Log.Reason(fmt.Errorf("tombstone contained object that is not a domain %#v", obj)).Error("Failed to process delete notification")
			return
		}
		return
	}
```

Reply:
i added logic to simply return early if the kubevirt object can't be cast. there's nothing for this control loop to do if the operator cr is deleted, so i don't see any value in retrieving the tombstone.

Reply:
Ack.